src/ucp/core/ucp_ep.c

```diff
@@ -834,60 +840,92 @@ int ucp_ep_config_is_equal(const ucp_ep_config_key_t *key1,
-        md_reg_growth   = md_attr->reg_cost.growth;
-        md_reg_overhead = md_attr->reg_cost.overhead;
-    } else {
-        md_reg_growth   = 0;
-        md_reg_overhead = 0;
+    ucp_context_h context         = worker->context;
+    ucp_md_map_t md_map           = 0;
+    ucp_ep_thresh_params_t params = {0};
+    ucp_lane_index_t lane;
+    ucp_rsc_index_t rsc_index;
+    ucp_md_index_t md_index;
+    uct_md_attr_t *md_attr;
+    int i;
+
+    for (i = 0; lanes[i] != UCP_NULL_LANE; i++) {
+        lane      = lanes[i];
+        rsc_index = config->key.lanes[lane].rsc_index;
+        md_index  = config->md_index[lane];
+
+        if (!(md_map & UCS_BIT(md_index))) {
+            md_map |= UCS_BIT(md_index);
+            md_attr = &context->tl_mds[md_index].attr;
+            if (md_attr->cap.flags & UCT_MD_FLAG_REG) {
+                params.reg_growth   += md_attr->reg_cost.growth;
+                params.reg_overhead += md_attr->reg_cost.overhead;
+            }
+        }
+        params.bw += worker->ifaces[rsc_index].attr.bandwidth;
+    }
+
+    *params_p = params;
+}
+
+static size_t ucp_ep_config_calc_rndv_thresh(ucp_worker_t *worker,
+                                             const ucp_ep_config_t *config,
+                                             const ucp_lane_index_t *eager_lanes,
+                                             const ucp_lane_index_t *rndv_lanes,
+                                             int recv_reg_cost)
+{
+    ucp_context_h context = worker->context;
+    double diff_percent   = 1.0 - context->config.ext.rndv_perf_diff / 100.0;
+    ucp_ep_thresh_params_t eager;
+    ucp_ep_thresh_params_t rndv;
+    double eager_latency;
+    double rndv_latency;
+    double eager_overhead;
+    double rndv_overhead;
+    double numerator, denumerator;
+    ucp_rsc_index_t rsc_index;
+    uct_iface_attr_t *iface_attr;
+
+    /* All formulas and descriptions are listed at
+     * https://github.com/openucx/ucx/wiki/Rendezvous-Protocol-threshold-for-multilane-mode */
+
+    ucp_ep_config_calc_thresh_params(worker, config, eager_lanes, &eager);
+    ucp_ep_config_calc_thresh_params(worker, config, rndv_lanes, &rndv);
+
+    if (!eager.bw || !rndv.bw) {
+        goto fallback;
     }
 
-    numerator = diff_percent * ((4 * ucp_tl_iface_latency(context, iface_attr)) +
-                (3 * iface_attr->overhead) +
-                (md_reg_overhead * (1 + recv_reg_cost))) -
-                 md_reg_overhead - iface_attr->overhead;
+    rsc_index  = config->key.lanes[eager_lanes[0]].rsc_index;
+    iface_attr = &worker->ifaces[rsc_index].attr;
+
+    /* we always using AM lane (eager_lanes[0]) for signal message (RTS/RTR/ATS),
+     * that is why eager lateny/overhead is equal to rndv latency/overhead */
+    eager_latency  = ucp_tl_iface_latency(context, iface_attr);
+    rndv_latency   = eager_latency;
+    eager_overhead = iface_attr->overhead;
+    rndv_overhead  = eager_overhead;
```

Comment:
i think it's better to calculate overhead as part of ucp_ep_config_calc_thresh_params() - even though it's the same

Reply:
ok, done